### PR TITLE
[Term Entry] PyTorch Tensor Operations .index_reduce()

### LINKED ISSUE #5616: [Term Entry] PyTorch Tensor Operations .index_reduce()
### Reason/inspiration (optional)

We would like a new entry on the `.index_reduce()` term under PyTorch. The entry should go in a new file under `docs/content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md`.

The entry should include:

- An introduction to the concept
- A `Syntax` section that provides the syntax for the concept
- An `Example` section that provides an example demonstrating the concept in use

Please refer to the [term entry template](https://github.com/Codecademy/docs/blob/main/documentation/term-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md) and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Term for Existing Concept Entry

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by mamtawardhani**
### Description
Added a new Term Entry for PyTorch Tensor Operations `.index_reduce()`

### Issue Solved
Closes #5616 

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [ ] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

File: content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+In PyTorch, **`.index_reduce()`** performs a reduction operation (such as sum, prod, or mean) on a [tensor](https://www.codecademy.com/resources/docs/pytorch/tensors) along a specified dimension, using an index tensor to map input elements to positions in the output tensor, effectively aggregating values with the same index.
+
+## Syntax
+
+```pseudo
+Tensor.index_reduce_(dim, index, source, reduce, *, include_self=True)
+```
+
+- `dim`: The dimension along which the reduction is performed. It specifies which axis of the tensor is being reduced.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:25]: There is repetition in this explanation. We can rewrite it simply as follows:

```suggestion
- `dim`: The axis of the tensor along which the reduction is performed.
```

File: content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+# Define the target tensor
+target = torch.zeros(5)
+
+# Source tensor
+source = torch.tensor([1.0, 2.0, 3.0, 4.0])
+
+# Indices mapping source to target
+index = torch.tensor([0, 1, 0, 1])
+
+# Perform in-place reduction using 'mean'
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:53]: The parameter `dim=0` is used in the example, which is fine, but it may not be immediately clear why dimension 0 was chosen. For users unfamiliar with PyTorch tensor dimensions, a brief explanation would help.

```suggestion
# Perform in-place reduction using 'mean' along the 0th dimension (rows)
```

File: content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  - 'Data Structures'
+  - 'Functions'
+  - 'Index'
+  - 'Values'
+CatalogContent:
+  - 'intro-to-py-torch-and-neural-networks'
+  - 'paths/computer-science'
+---
+
+In PyTorch, **`.index_reduce()`** performs a reduction operation (such as sum, prod, or mean) on a [tensor](https://www.codecademy.com/resources/docs/pytorch/tensors) along a specified dimension, using an index tensor to map input elements to positions in the output tensor, effectively aggregating values with the same index.
```
[the remark this answers is not in the document]
Comment (PragatiVerma18) [reply content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:17]: We have not explained why we are using `.index_reduce()` in some places and why `.index_reduce_()` in others; this is confusing. Please add an explanation somewhere.

File: content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import torch
+
+# Define the target tensor
+target = torch.zeros(5)
+
+# Source tensor
+source = torch.tensor([1.0, 2.0, 3.0, 4.0])
+
+# Indices mapping source to target
+index = torch.tensor([0, 1, 0, 1])
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:51]: It's important to highlight the type requirement for the `index` tensor, as PyTorch typically requires integer types for indexing. The operation may fail if the `index` tensor is not of an integer type.

```suggestion
index = torch.tensor([0, 1, 0, 1], dtype=torch.long)  # Ensure index tensor is of type 'long'
```

File: content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+## Example
+
+The following example demonstrates the usage of the `.index_reduce()` method:
+
+```py
+import torch
+
+# Define the target tensor
+target = torch.zeros(5)
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/tensor-operations/terms/index-reduce/index-reduce.md:45]: The input tensors (`target`, `source`, and `index`) have different shapes. The `index` tensor has 4 elements, and the `source` tensor has 4 elements, but `target` is initialized with 5 elements. 

This may cause confusion because the `target` tensor should match the number of output locations determined by the `index` tensor.

Ensure that the `target` tensor matches the number of unique indices in `index`. If `index` contains values mapping to positions 0 and 1, for example, the `target` tensor should have at least 2 elements.

```suggestion
# Define the target tensor
target = torch.zeros(2)
```

Please update output accordingly.

**Review by PragatiVerma18 — CHANGES_REQUESTED** [review]
Hey @mamtawardhani, thanks for adding this entry. I have some suggestions, please check and do the required fixes. Let me know if you have any questions. Happy contributing! 🥳

**Comment by mamtawardhani** [comment]
Hi @PragatiVerma18, thanks for the feedback - I've made the changes, also I've updated the entry to be `.index_reduce_()` instead of `.index_reduce(), as we are working on tensor operations. I've updated the issue as well

**Review by PragatiVerma18 — APPROVED** [review]
Hey @mamtawardhani, thanks for doing the changes so quickly. This looks good to be merged! 🚀
